rmltraintfbbox/rmltraintfbbox/core.py

```diff
@@ -119,78 +122,135 @@ def train(ctx: click.Context, train: TrainInput):
+
+    train_input_iterator = iter(train_input)
+
+    global_step = tf.Variable(
+        0, trainable=False, dtype=tf.int64, name='global_step',
+        aggregation=tf.VariableAggregation.ONLY_FIRST_REPLICA)
+    optimizer, (learning_rate,) = optimizer_builder.build(
+        train_config.optimizer, global_step=global_step)
+
+    if callable(learning_rate):
+      learning_rate_fn = learning_rate
+    else:
+      learning_rate_fn = lambda: learning_rate
+
+    #restore from checkpoint
+    load_fine_tune_checkpoint(detection_model, train_config.fine_tune_checkpoint, 
+                                        train_config.fine_tune_checkpoint_type,
+                                        train_config.fine_tune_checkpoint_version,
+                                        train_input, 
+                                        train_config.unpad_groundtruth_tensors)
+
+    @tf.function
+    def train_step(detection_model, train_input_iterator, optimizer, learning_rate_fn, global_step):
+
+        features, labels = train_input_iterator.next()
+        loss = model_lib_v2.eager_train_step(detection_model, features,
+                            labels, train_config.unpad_groundtruth_tensors,
+                            optimizer, learning_rate_fn(),
+                            add_regularization_loss=True,
+                            clip_gradients_value=None,
+                            global_step=global_step)
+        global_step.assign_add(1)
+        return loss
+
+    #@tf.function
+    def evaluate(detection_model, configs, eval_input, global_step):
+
+        text_trap = io.StringIO()
+        sys.stdout = text_trap
+        sys.stderr = text_trap
+        
+        metrics = model_lib_v2.eager_eval_loop(detection_model, configs, eval_input, global_step=global_step)
+        
+        sys.stdout = sys.__stdout__
+        sys.stderr = sys.__stderr__
+
+        click.echo(f'Evaluation loss: {metrics["Loss/total_loss"]}, Evaluation mAP: {metrics["DetectionBoxes_Precision/mAP"]}')
+
+        return metrics
+
+    checkpoint = tf.train.Checkpoint(optimizer=optimizer, model=detection_model)
+    manager = tf.train.CheckpointManager(checkpoint, directory=model_dir, max_to_keep=5)
+      
     with ExitStack() as stack:
         if comet:
             stack.enter_context(experiment.train())
         click.echo('Training model...')
-        tf.estimator.train_and_evaluate(estimator, train_spec, eval_specs[0])
-        click.echo('Training complete')
+
+        start = time.time()
+        #main training loop
+        for step in range(1, num_train_steps+1):
+
+            loss = train_step(detection_model, train_input_iterator, optimizer, learning_rate_fn, global_step)
+            
+            if step % 100 == 0:
+                print(f'Training loss at step {step}: {loss}')
+                manager.save()
+
+            if step % 500 == 0:
```

Comment:
Looking at a full example run (https://www.comet.ml/seeker-rd/bounding-box/bc7a6607447a4d2fa600c36464366845?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step), it seems like all of the train metrics are getting logged properly. I'm guessing this is because comet automatically injects itself into any Tensorboard logging. However, I just noticed that it's actually the eval metrics that are missing, which are arguably more important -- @dlfonseka any idea why?

Reply:
Ah, actually, ignore the above. I realized that the metrics being logged are actually the eval metrics, comet just appends `train_` because it is still in a train context. @cSchubes is correct, we should figure out how to log the train loss as well. You should also probably make sure that comet is in the correct context when logging metrics to avoid confusion.

Reply:
On the first point, I don't think that the frequency of logging/evaluation/checkpoint saving should be based on the total number of train steps -- every fixed amount of steps is the right way to do it. However, it would be nice if these numbers were configurable, since 100/500 might be too frequent or infrequent depending on how fast your hardware is.

Reply:
Steps between train/eval logging to comet now configurable in the plugin specific options of the config files. The naming scheme of the comet logging is also fixed.